src/commands/tail/mod.rs

```diff
@@ -0,0 +1,220 @@
+use log::log_enabled;
+use log::Level::Info;
+use regex::Regex;
+use reqwest;
+use tokio;
+use tokio::process::Command;
+use tokio::runtime::Runtime as TokioRuntime;
+
+pub fn start_tail(target: &Target, user: &GlobalUser) -> Result<(), failure::Error> {
+    let mut runtime = TokioRuntime::new()?;
+    // Note that we use eprintln!() throughout this file; this is because we want any
+    // helpful output to not be mixed with actual log JSON output, so we use this macro
+    // to print messages to stderr instead of stdout (where log output is printed).
+    eprintln!(
+        "Setting up log streaming from Worker \"{}\" to Wrangler. This may take a few seconds...",
+        target.name
+    );
+
+    // TODO: get rid of this gross clone. Maybe just default to static lifetime.
+    runtime.block_on(run_cloudflared_start_server(target.clone(), user.clone()))
+}
+
+// TODO: handle sigint to tear down tunnel, http server
+async fn run_cloudflared_start_server(
+    target: Target,
+    user: GlobalUser,
+) -> Result<(), failure::Error> {
+    let res = tokio::try_join!(
+        start_log_collection_http_server(),
+        start_argo_tunnel(),
+        enable_tailing_start_heartbeat(&target, &user)
+    );
+
+    match res {
+        Ok(_) => Ok(()),
+        Err(e) => failure::bail!(e),
+    }
+}
+
+async fn start_log_collection_http_server() -> Result<(), failure::Error> {
+    // Start HTTP echo server that prints whatever is posted to it.
+    let addr = ([127, 0, 0, 1], 8080).into();
+
+    let service = make_service_fn(|_| async { Ok::<_, hyper::Error>(service_fn(print_logs)) });
+
+    let server = Server::bind(&addr).serve(service);
+
+    server.await?;
+
+    Ok(())
+}
+
+async fn start_argo_tunnel() -> Result<(), failure::Error> {
+    // TODO: remove sleep!! Can maybe use channel to signal from http server thread to argo tunnel
+    // thread that the server is ready on port 8080 and prepared for the cloudflared CLI to open an
+    // Argo Tunnel to it.
+    thread::sleep(Duration::from_secs(5));
+
+    let tool_name = PathBuf::from("cloudflared");
+    // TODO: Finally get cloudflared release binaries distributed on GitHub so we could simply uncomment
+    // the line below.
+    // let binary_path = install::install(tool_name, "cloudflare")?.binary(tool_name)?;
+
+    // TODO: allow user to pass in their own ports in case ports 8080 (used by cloudflared)
+    // and 8081 (used by cloudflared metrics) are both already being used.
+    let args = ["tunnel", "--metrics", "localhost:8081"];
+
+    let mut command = command(&args, &tool_name);
+    let command_name = format!("{:?}", command);
+
+    let status = command
+        .kill_on_drop(true)
+        .spawn()
+        .expect(&format!("{} failed to spawn", command_name))
+        .await?;
+
+    if !status.success() {
+        failure::bail!(
+            "tried running command:\n{}\nexited with {}",
+            command_name.replace("\"", ""),
```

Comment:
is there a `command_name.display()` function?

Reply:
possibly; this is also a thing that may change.